crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2074 @@
+            // must be recorded in the changes.
+            // The Java code might or might not change `lastAddedSchemaId`, and does not record this change in `changes`.
+            // Thus, replaying the changes, would lead to a different result if a schema is added twice in unfavorable
+            // conditions.
+            // Here we do it differently, but a check from a Java maintainer would be nice.
+            if self.last_added_schema_id != Some(new_schema_id) {
+                self.changes.push(TableUpdate::AddSchema {
+                    last_column_id: Some(self.metadata.last_column_id),
+                    schema: schema.clone(),
+                });
+                self.last_added_schema_id = Some(new_schema_id);
+            }
+
+            return self;
+        }
+
+        // New schemas might contain only old columns. In this case last_column_id should not be
+        // reduced.
+        self.metadata.last_column_id =
+            std::cmp::max(self.metadata.last_column_id, schema.highest_field_id());
+
+        // Set schema-id
+        let schema = match new_schema_id == schema.schema_id() {
+            true => schema,
+            false => schema.with_schema_id(new_schema_id),
+        };
+
+        self.metadata
+            .schemas
+            .insert(new_schema_id, schema.clone().into());
+
+        self.changes.push(TableUpdate::AddSchema {
+            schema,
+            last_column_id: Some(self.metadata.last_column_id),
+        });
+
+        self.last_added_schema_id = Some(new_schema_id);
+
+        self
+    }
+
+    /// Set the current schema id.
+    ///
+    /// Errors:
+    /// - provided `schema_id` is -1 but no schema has been added via `add_schema`.
+    /// - No schema with the provided `schema_id` exists.
+    pub fn set_current_schema(mut self, mut schema_id: i32) -> Result<Self> {
+        if schema_id == Self::LAST_ADDED {
+            schema_id = self.last_added_schema_id.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    "Cannot set current schema to last added schema: no schema has been added.",
+                )
+            })?;
+        };
+        let schema_id = schema_id; // Make immutable
+
+        if schema_id == self.metadata.current_schema_id {
+            return Ok(self);
+        }
+
+        let _schema = self.metadata.schemas.get(&schema_id).ok_or_else(|| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Cannot set current schema to unknown schema with id: '{}'",
+                    schema_id
+                ),
+            )
+        })?;
+
+        // Old partition specs and sort-orders should be preserved even if they are not compatible with the new schema,
+        // so that older metadata can still be interpreted.
```

Comment:
Incompatible changes should not be made to the schema

Reply:
Oh, but then I think we should also re-check java. It is possible to have non-compatible changes due to:
https://github.com/apache/iceberg/blob/166edc7298825321f677e1e70cf88d7249e8035c/core/src/main/java/org/apache/iceberg/TableMetadata.java#L740
Its using an unchecked build there - so even if it is not compatible, it can build.
I copied the comment more or less from java. The java comment is:

  // build without validation because the schema may have changed in a way that makes this spec
  // invalid. the spec
  // should still be preserved so that older metadata can be interpreted.